### DISCUSSION TITLE
Create and instantiate a class

### DISCUSSION THREAD

**Discussion by gregghz** [scala/scala3#15125] [Metaprogramming] [answered]
```
object Macro {
  inline def create[T]: T = ${impl}
  private def impl[T: Type](using Quotes): Expr[T] = {
    val t = Type.of[T]

    '{
      final class SomeRandomName extends $t {
        def someMethod(): String = ""
      }
      new SomeRandomName()
    }    
  }
}
```

This fails with `T is not a class type`.

I'm hoping to write this macro such that this code:

```
val x = Macro.create[SomeClass]
```

expands into (or something equivalent):

```
val x = {
  final class SomeRandomName extends SomeClass {
    def someMethod(): String = ""
  }
  new SomeRandomName()
}
```

I've been struggling to know how to use the reflect API and build an AST from that mostly because the results of term.show are different than how you actually use the API (for example, when printing, TypeDef shows two parameters but the only constructors that exist take one or three parameters -- and for copy, it's unclear to me what the `original` tree is supposed to be and what happens when you "copy").

**Comment by bishabosha** [comment, marked as answer]
I suggest that you look at the examples in this PR:
https://github.com/lampepfl/dotty/pull/14124/ but you will be using experimental API (released in Scala 3.1.3-RC2)

**Reply by gregghz** [reply]
do you happen to know how to use an `@experimental` method (like `Symbol.newClass`) inside an `inline` method?

https://scastie.scala-lang.org/CCUeignRSJOnYEBa8aJTWg

I filed a bug for this but I think it might be intentional? https://github.com/lampepfl/dotty/issues/15133

**Reply by nicolasstucki** [reply]
A simpler approach is to use a nightly build of the compiler. There you will not need to use the `@experimental` annotation in your code.

**Reply by nicolasstucki** [reply]
Using the current nightly will not work due to https://github.com/lampepfl/dotty/issues/15133.  The nightly published after https://github.com/lampepfl/dotty/pull/15147 is merged will work.

**Reply by gregghz** [reply]
bummer, looks like the next RC got cut without this being merged :(
